classes/boardmanager.php

```diff
@@ -0,0 +1,1041 @@
+            'timecreated' => time(),
+            'timemodified' => time(),
+            'sequence' => '',
+        ];
+        $data = array_merge($defaults, $data, $defaultsfixed);
+
+        $data['id'] = $DB->insert_record('kanban_card', $data);
+        $data['assignees'] = [];
+
+        $column = $DB->get_record('kanban_column', ['id' => $columnid]);
+
+        $update = [
+            'id' => $columnid,
+            'sequence' => helper::sequence_add_after($column->sequence, $aftercard, $data['id']),
+            'timemodified' => time()
+        ];
+        $DB->update_record('kanban_column', $update);
+
+        // Users can always edit cards they created.
+        $data['canedit'] = true;
+        $data['columnname'] = $column->title;
+
+        $this->formatter->put('cards', $data);
+        $this->formatter->put('columns', $update);
+        $this->write_history('added', constants::MOD_KANBAN_CARD, $data, $columnid, $data['id']);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_COLUMN, $update['timemodified']);
+        helper::update_cached_timestamp($this->board->id, constants::MOD_KANBAN_CARD, $update['timemodified']);
+
+        return $data['id'];
+    }
+
+    /**
+     * Moves a column.
+     *
+     * @param int $columnid Id of the column to move
+     * @param int $aftercol Id of the (future) column before (0 means to move at the leftmost position)
+     * @return void
+     */
+    public function move_column(int $columnid, int $aftercol) {
+        global $DB;
+        $column = $DB->get_record('kanban_column', ['id' => $columnid]);
+        if (!$this->board->locked && !$column->locked) {
+            $update = [
+                'id' => $this->board->id,
+                'sequence' => helper::sequence_move_after($this->board->sequence, $aftercol, $columnid),
+                'timemodified' => time(),
+            ];
+            $DB->update_record('kanban_board', $update);
+            helper::update_cached_board($update['id']);
+            $this->formatter->put('board', $update);
+        }
+    }
+
+    /**
+     * Moves a card.
+     *
+     * @param int $cardid Id of the card to move
+     * @param int $aftercard If of the card to move after (0 means move to top of the column)
+     * @param int $columnid Id of the column to move to (if 0, use current column)
+     * @return void
+     */
+    public function move_card(int $cardid, int $aftercard, int $columnid = 0) {
+        global $DB, $USER;
+        $card = $this->get_card($cardid);
+        if (empty($columnid)) {
+            $columnid = $card->kanban_column;
+        }
+
+        $sourcecolumn = $DB->get_record('kanban_column', ['id' => $card->kanban_column]);
+
+        if ($card->kanban_column == $columnid) {
+            $update = [
+                'id' => $columnid,
+                'sequence' => helper::sequence_move_after($sourcecolumn->sequence, $aftercard, $cardid),
+                'timemodified' => time()
```

Comment:
again, maybe think of the same timestamp for the whole function

Reply:
I think there is no special need for having the same timestamp as it has no effect on updating the board.